tools/sanity_check.py

```diff
@@ -0,0 +1,193 @@
+# pylint: disable=invalid-name
+# pylint: disable=line-too-long
+# pylint: disable=missing-function-docstring
+"""A script to perform sanity checks for Bazel modules
+
+Sanity checks performed are:
+  - Verify versions in metadata.json matches existing versions
+  - Verify the source archive URL matches the source repositories
+  - Verify the source archive URL is stable
+  - Verify if the presubmit.yml file matches the previous version
+    - If not, we should require BCR maintainer review.
+"""
+
+import argparse
+import sys
+
+from enum import Enum
+from difflib import unified_diff
+
+from registry import RegistryClient
+from registry import Version
+from verify_stable_archives import UrlStability
+from verify_stable_archives import verify_stable_archive
+
+
+class SanityCheckResult(Enum):
+  ALL_GOOD = 1
+  NEED_BCR_MAINTAINER_REVIEW = 2
+  FAILED = 3
+
+RED = "\x1b[31m"
+GREEN = "\x1b[32m"
+YELLOW = "\x1b[33m"
+RESET = "\x1b[0m"
+
+COLOR = {
+  SanityCheckResult.ALL_GOOD: GREEN,
+  SanityCheckResult.NEED_BCR_MAINTAINER_REVIEW: YELLOW,
+  SanityCheckResult.FAILED: RED,
+}
+
+def print_collapsed_group(name):
+    print("\n\n--- {0}\n\n".format(name))
+
+def print_expanded_group(name):
+    print("\n\n+++ {0}\n\n".format(name))
+
+def parse_module_versions(registry, check_all, inputs):
+  """Parse module versions to be checked from input."""
+  if check_all:
+    return registry.get_all_module_versions()
+  result = []
+  for s in inputs:
+    if "@" in s:
+      name, version = s.split("@")
+      result.append((name, version))
+    else:
+      result.extend(registry.get_module_versions(s))
+  return result
+
+def print_check_result(result):
+  for code, message in result:
+    color = COLOR[code]
+    print(f"{color}{code}{RESET}: {message}\n")
+
+def sanity_check(registry, module_name, version):
+  print_collapsed_group(f"Checking {module_name}@{version}")
+
+  check_results = []
+
+  # Check the version is recorded in metadata.json and the directory entry exists.
+  versions = registry.get_metadata(module_name)["versions"]
+  if version not in versions:
+    check_results.append((SanityCheckResult.FAILED, f"Version {version} is not recorded in {module_name}'s metadata.json file."))
+  if not registry.contains(module_name, version):
+    check_results.append((SanityCheckResult.FAILED, f"{module_name}@{version} doesn't exist."))
+  if check_results:
+    print_check_result(check_results)
+    return check_results
+
+  # Verify the source archive URL matches the github repo. For now, we only support github repositories check.
+  source_url = registry.get_source(module_name, version)["url"]
+  source_repositories = registry.get_metadata(module_name).get("repository", [])
+  matched = False
+  for source_repository in source_repositories:
+    if matched:
+      break
+    repo_type, repo_path = source_repository.split(":")
+    if repo_type == "github":
+      matched = source_url.startswith("https://github.com/" + repo_path)
```

Comment:
add an extra '/' at the end so that "github.com/foo/bar_bad" doesn't match "github:foo/bar"

Reply:
Nice catch!